File: maven-plugin/src/test/java/MavenUtilsTest.java
import org.apache.maven.plugin.testing.MojoRule;
import org.apache.maven.plugin.testing.WithoutMojo;
import org.apache.maven.project.MavenProject;
import org.junit.Rule;
import org.junit.Test;
import static org.junit.Assert.*;
import java.io.File;

/**
 * Unit tests for MavenUtils utility methods using Maven Plugin Testing Harness.
 */
public class MavenUtilsTest {

    @Rule
    public MojoRule rule = new MojoRule() {
        @Override
        protected void before() throws Throwable {
        }

        @Override
        protected void after() {
        }
    };

    /**
     * Test phase inference from goal names with real Maven project
     */
    @Test
    public void testInferPhaseFromGoal() throws Exception {
        File pom = new File("target/test-classes/unit/basic-test");
        assertNotNull(pom);
        assertTrue(pom.exists());

        NxAnalyzerMojo mojo = (NxAnalyzerMojo) rule.lookupConfiguredMojo(pom, "analyze");
        assertNotNull(mojo);
        
        // Test with a real Maven project from the mojo
        MavenProject project = new MavenProject();
        
        // Get session from mojo
        org.apache.maven.execution.MavenSession session = (org.apache.maven.execution.MavenSession) rule.getVariableValueFromObject(mojo, "session");
        
        // Note: MavenUtils.inferPhaseFromGoal has been removed in favor of ExecutionPlanAnalysisService
        // This test now demonstrates the new approach
        ExecutionPlanAnalysisService analysisService = new ExecutionPlanAnalysisService(mojo.getLog(), false, null, session);
        
        // Test null handling
        assertNull("Null goal should return null", analysisService.findPhaseForGoal(project, null));
        assertNull("Empty goal should return null", analysisService.findPhaseForGoal(project, ""));
        
        // Test with null project should handle gracefully (may return null or throw exception depending on implementation)
        String result = analysisService.findPhaseForGoal(null, "compile");
        // This may be null or throw an exception - both are acceptable for null input
    }

    /**
     * Test target name generation (moved to ExecutionPlanAnalysisService)
     */
    @Test
    @WithoutMojo
    public void testGetTargetName() {
        assertEquals("maven-compiler:compile", ExecutionPlanAnalysisService.getTargetName("maven-compiler-plugin", "compile"));
        assertEquals("maven-surefire:test", ExecutionPlanAnalysisService.getTargetName("maven-surefire-plugin", "test"));
        assertEquals("maven-jar:jar", ExecutionPlanAnalysisService.getTargetName("maven-jar-plugin", "jar"));
    }

    /**
     * Test goal extraction from target names (moved to ExecutionPlanAnalysisService)
     */
    @Test
    @WithoutMojo
    public void testExtractGoalFromTargetName() {
        assertEquals("compile", ExecutionPlanAnalysisService.extractGoalFromTargetName("maven-compiler:compile"));
        assertEquals("test", ExecutionPlanAnalysisService.extractGoalFromTargetName("maven-surefire:test"));
        assertEquals("jar", ExecutionPlanAnalysisService.extractGoalFromTargetName("maven-jar:jar"));
        
        // Test edge cases
        assertEquals("complex-goal", ExecutionPlanAnalysisService.extractGoalFromTargetName("plugin:complex-goal"));
        assertEquals("single", ExecutionPlanAnalysisService.extractGoalFromTargetName("single"));
    }

    /**
     * Test project key formatting (static utility method)
     */
    @Test
    @WithoutMojo
    public void testFormatProjectKey() {
        MavenProject project = new MavenProject();
        project.setGroupId("com.example");
        project.setArtifactId("test-project");
        
        assertEquals("com.example:test-project", MavenUtils.formatProjectKey(project));
    }

    /**
     * Test plugin name normalization (new method in ExecutionPlanAnalysisService)
     */
    @Test
    @WithoutMojo
    public void testNormalizePluginName() {
        assertEquals("maven-compiler", ExecutionPlanAnalysisService.normalizePluginName("maven-compiler-plugin"));
        assertEquals("maven-surefire", ExecutionPlanAnalysisService.normalizePluginName("maven-surefire-plugin"));
        assertEquals("quarkus", ExecutionPlanAnalysisService.normalizePluginName("quarkus-maven-plugin"));
        assertEquals("simple", ExecutionPlanAnalysisService.normalizePluginName("simple-plugin"));
        assertEquals("test", ExecutionPlanAnalysisService.normalizePluginName("test"));
        assertNull(ExecutionPlanAnalysisService.normalizePluginName(null));
    }

    /**
     * Test common goals for plugins (new method in ExecutionPlanAnalysisService)
     */
    @Test
    @WithoutMojo
    public void testGetCommonGoalsForPlugin() {
        // Test compiler plugin
        java.util.List<String> compilerGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("maven-compiler-plugin");
        assertEquals(2, compilerGoals.size());
        assertTrue(compilerGoals.contains("compile"));
        assertTrue(compilerGoals.contains("testCompile"));

        // Test surefire plugin
        java.util.List<String> surefireGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("maven-surefire-plugin");
        assertEquals(1, surefireGoals.size());
        assertTrue(surefireGoals.contains("test"));

        // Test quarkus plugin
        java.util.List<String> quarkusGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("quarkus-maven-plugin");
        assertEquals(2, quarkusGoals.size());
        assertTrue(quarkusGoals.contains("dev"));
        assertTrue(quarkusGoals.contains("build"));

        // Test unknown plugin
        java.util.List<String> unknownGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("unknown-plugin");
        assertEquals(0, unknownGoals.size());

        // Test null input
        java.util.List<String> nullGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin(null);
        assertEquals(0, nullGoals.size());
    }
}

File: maven-plugin/src/test/kotlin/MavenUtilsTest.kt
import org.apache.maven.execution.MavenSession
import org.apache.maven.lifecycle.DefaultLifecycles
import org.apache.maven.lifecycle.LifecycleExecutor
import org.apache.maven.plugin.logging.Log
import org.apache.maven.plugin.testing.MojoRule
import org.apache.maven.project.MavenProject
import org.junit.Rule
import org.junit.Test
import java.io.File
import kotlin.test.*

/**
 * Unit tests for MavenUtils utility methods using Maven Plugin Testing Harness.
 */
class MavenUtilsTest {

    @get:Rule
    val rule = object : MojoRule() {
        override fun before() {
        }

        override fun after() {
        }
    }

    /**
     * Test phase inference from goal names with real Maven project
     */
    @Test
    fun testInferPhaseFromGoal() {
        val pom = File("target/test-classes/unit/basic-test")
        assertNotNull(pom)
        assertTrue(pom.exists())

        val mojo = rule.lookupConfiguredMojo(pom, "analyze") as NxAnalyzerMojo
        assertNotNull(mojo)
        
        // Test with a real Maven project from the mojo
        val project = MavenProject()
        
        // Get session from mojo
        val session = rule.getVariableValueFromObject(mojo, "session") as org.apache.maven.execution.MavenSession?
        
        // Note: MavenUtils.inferPhaseFromGoal has been removed in favor of ExecutionPlanAnalysisService
        // This test now demonstrates the new approach
        val defaultLifecycles = rule.getVariableValueFromObject(mojo, "defaultLifecycles") as DefaultLifecycles?
        val lifecycleExecutor = rule.getVariableValueFromObject(mojo, "lifecycleExecutor") as LifecycleExecutor
        val analysisService = ExecutionPlanAnalysisService(mojo.log, false, lifecycleExecutor, session!!, defaultLifecycles!!)
        
        // Test null handling
        assertNull(analysisService.findPhaseForGoal(project, null), "Null goal should return null")
        
        // Test empty goal handling
        assertNull(analysisService.findPhaseForGoal(project, ""), "Empty goal should return null")
        
        // Test with common goals - may return null in test environment, which is acceptable
        val compilePhase = analysisService.findPhaseForGoal(project, "compile")
        val testPhase = analysisService.findPhaseForGoal(project, "test")
        
        // These might be null in test environment - that's acceptable
        // The important thing is no exceptions are thrown
    }

    /**
     * Test target name generation utilities
     */
    @Test
    fun testTargetNameUtilities() {
        // Test getTargetName
        val targetName1 = ExecutionPlanAnalysisService.getTargetName("maven-compiler-plugin", "compile")
        assertEquals("maven-compiler:compile", targetName1, "Should generate correct target name")
        
        val targetName2 = ExecutionPlanAnalysisService.getTargetName("maven-surefire-plugin", "test")
        assertEquals("maven-surefire:test", targetName2, "Should generate correct target name")
        
        // Test extractGoalFromTargetName
        val goal1 = ExecutionPlanAnalysisService.extractGoalFromTargetName("compiler:compile")
        assertEquals("compile", goal1, "Should extract goal correctly")
        
        val goal2 = ExecutionPlanAnalysisService.extractGoalFromTargetName("maven-surefire-plugin:test")
        assertEquals("test", goal2, "Should extract goal correctly")
        
        // Test null handling
        val nullGoal = ExecutionPlanAnalysisService.extractGoalFromTargetName(null)
        assertNull(nullGoal, "Should handle null input")
        
        // Test without colon
        val simpleGoal = ExecutionPlanAnalysisService.extractGoalFromTargetName("compile")
        assertEquals("compile", simpleGoal, "Should return input if no colon")
    }

    /**
     * Test plugin name normalization
     */
    @Test
    fun testNormalizePluginName() {
        // Test removing maven-plugin suffix
        val normalized1 = ExecutionPlanAnalysisService.normalizePluginName("maven-compiler-plugin")
        assertEquals("maven-compiler", normalized1, "Should remove -maven-plugin suffix")
        
        // Test removing plugin suffix
        val normalized2 = ExecutionPlanAnalysisService.normalizePluginName("kotlin-plugin")
        assertEquals("kotlin", normalized2, "Should remove -plugin suffix")
        
        // Test with no suffix
        val normalized3 = ExecutionPlanAnalysisService.normalizePluginName("simple")
        assertEquals("simple", normalized3, "Should return input if no suffix")
        
        // Test null handling
        val nullResult = ExecutionPlanAnalysisService.normalizePluginName(null)
        assertNull(nullResult, "Should handle null input")
    }

    /**
     * Test common goals for plugins
     */
    @Test
    fun testGetCommonGoalsForPlugin() {
        // Test compiler plugin
        val compilerGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("maven-compiler-plugin")
        assertTrue(compilerGoals.contains("compile"), "Should contain compile goal")
        assertTrue(compilerGoals.contains("testCompile"), "Should contain testCompile goal")
        
        // Test surefire plugin
        val surefireGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("maven-surefire-plugin")
        assertTrue(surefireGoals.contains("test"), "Should contain test goal")
        
        // Test quarkus plugin
        val quarkusGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("quarkus-maven-plugin")
        assertTrue(quarkusGoals.contains("dev"), "Should contain dev goal")
        assertTrue(quarkusGoals.contains("build"), "Should contain build goal")
        
        // Test spring boot plugin
        val springBootGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("spring-boot-maven-plugin")
        assertTrue(springBootGoals.contains("run"), "Should contain run goal")
        assertTrue(springBootGoals.contains("repackage"), "Should contain repackage goal")
        
        // Test unknown plugin
        val unknownGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin("unknown-plugin")
        assertTrue(unknownGoals.isEmpty(), "Should return empty list for unknown plugin")
        
        // Test null handling
        val nullGoals = ExecutionPlanAnalysisService.getCommonGoalsForPlugin(null)
        assertTrue(nullGoals.isEmpty(), "Should return empty list for null input")
    }

    /**
     * Test goal outputs (simplified implementation)
     */
    @Test
    fun testGetGoalOutputs() {
        val pom = File("target/test-classes/unit/basic-test")
        assertTrue(pom.exists())

        val mojo = rule.lookupConfiguredMojo(pom, "analyze") as NxAnalyzerMojo
        val session = rule.getVariableValueFromObject(mojo, "session") as org.apache.maven.execution.MavenSession?
        val defaultLifecycles = rule.getVariableValueFromObject(mojo, "defaultLifecycles") as DefaultLifecycles?
        val lifecycleExecutor = rule.getVariableValueFromObject(mojo, "lifecycleExecutor") as LifecycleExecutor
        val analysisService = ExecutionPlanAnalysisService(mojo.log, false, lifecycleExecutor, session!!, defaultLifecycles!!)
        
        val project = MavenProject()
        
        // Test goal outputs - current implementation returns empty list
        val outputs = analysisService.getGoalOutputs("compile", "/project/root", project)
        assertNotNull(outputs, "Outputs should not be null")
        assertTrue(outputs.isEmpty(), "Current implementation returns empty list")
    }
}